### DISCUSSION TITLE
ensureIndexes

### DISCUSSION THREAD

**Discussion by rubiin** [mikro-orm/mikro-orm#6024] [Q&A] [answered]
When I set ensureIndexes : true , the orm always tries to create the index even if they exists and logs an error. Is this the actual behaviour?

**Comment by B4nan** [comment]
how do you define that index?

**Reply by rubiin** [reply]
```ts
import { Entity, Property } from "@mikro-orm/core";

@Entity()
export class Job {
  @Property({ index: true })
  title!: string;

  @Property({ index: true })
  code!: string;

  @Property({ length: 100_000 })
  summary!: string;

  @Property({ length: 100_000 })
  description!: string;

  @Property()
  location!: string;

  @Property({ index: true })
  isPublished = false;
}

```

On the property decorator like this

**Reply by rubiin** [reply]
@B4nan https://github.com/rubiin/reproduction heres a repro
Also a question , would the orm throw error if we do
```ts
await orm.schema.ensureIndexes();
await orm.schema.ensureIndexes();
```

**Reply by B4nan** [reply]
so its index on primary key? i dont think you need that, i doubt it would make any difference, if so, mongo is even weirder than i though :]

**Reply by rubiin** [reply]
Yeah that is true. Mongo behaves in some weird ways for some scenarios

**Reply by B4nan** [reply]
what is true? i just told you the repro is doing something that makes no sense, you don't need an index on a PK in the first place, at least i doubt it

edit: i mean primary key is almost certainly indexed automatically, just like in every other database.

**Reply by rubiin** [reply]
removed index on _id 

```ts
@Entity()
class User {

  @PrimaryKey()
  _id: ObjectId;

  @SerializedPrimaryKey()
  id!: string; // won't be saved in the database

  @Property()
  name: string;

  @Property({ unique: true ,index:true})
  email: string;
}
```

this still throws error

**Reply by B4nan** [reply]
eh, another "why would you do this" moment. a unique constraint doesn't need an explicit index.

**Reply by rubiin** [reply]
That was just an example of syntax I was using , ignore the constraint

**Reply by B4nan** [reply]
well, but we have tests for this, its not like its broken completely

e.g. https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/schema-generator/SchemaGenerator.mongo.test.ts

**Reply by B4nan** [reply]
i checked your repro, and it works just fine if you don't try to create both unique constraint and index on the same property. which i just told you is not needed, at least i doubt it.

**Reply by B4nan** [reply, marked as answer]
i think the problem is that both unique and index are technically indexes (that's why i keep telling that what you are doing makes no sense). 

i can try to fix this, but the fix will be just about ignoring your `index: true` which is not needed - the `unique: true` already creates (unique) index. that's why you see errors, the "normal" index is not possible to create since there is another one for the `unique: true`. in other words, just don't do this and you will be fine, having `unique: true` is enough.

**Reply by B4nan** [reply]
btw this is technically what is happening if you use both index and unique options:

```ts
await orm.em.getCollection(User).createIndex({ email: 1 });
await orm.em.getCollection(User).createIndex({ email: 1 }, { unique: true });
```

and if you would run that, you get a better understanding what is happening, this fails with

```
An existing index has the same name as the requested index. When index names are not specified, they are auto generated and can cause conflicts. Please refer to our documentation. Requested index: { v: 2, unique: true, key: { email: 1 }, name: "email_1" }, existing index: { v: 2, key: { email: 1 }, name: "email_1", sparse: false }
```

we really need better error propagation for this, and technically we should generate separate names for the indexes automatically so it can actually work (i would still say its just duplicating what you already have and will only result in slower inserts and bigger db size, but that's another topic)

**Reply by B4nan** [reply]
btw you can get around this by specifying the names explicitly:

````ts
@Property({ unique: 'email_u', index: 'email_i' })
email!: string;
```

so feel free to give that a try and verify if it makes any difference perf-wise or not.

**Reply by rubiin** [reply]
Sure thing

**Reply by B4nan** [reply]
This should make things clear hopefully faa2a0066f017082a508fe5ccd4a8d4d519ef280
